src/cli.ts

```diff
@@ -0,0 +1,254 @@
+
+	protected constructor(protected readonly umzug: Umzug<{}>, cliOptions: cli.ICommandLineActionOptions) {
+		super(cliOptions);
+	}
+
+	private static _defineParameters(action: cli.CommandLineAction) {
+		const verb = ApplyMigrationsAction.getVerb(action.actionName);
+
+		return {
+			to: action.defineStringParameter({
+				parameterLongName: '--to',
+				argumentName: 'NAME',
+				// prettier-ignore
+				description: `All migrations up to and including this one should be ${verb}. ${verb === 'reverted' ? 'Pass "0" to revert all.' : ''}`.trim(),
+			}),
+			step: action.defineIntegerParameter({
+				parameterLongName: '--step',
+				argumentName: 'COUNT',
+				// prettier-ignore
+				description: `Run this many migrations. If not specified, ${verb === 'reverted' ? 'one' : 'all'} will be ${verb}.`,
+			}),
+			name: action.defineStringListParameter({
+				parameterLongName: '--name',
+				argumentName: 'MIGRATION',
+				description: `Explicity declare migration name(s) to be ${verb}.`,
+			}),
+			rerun: action.defineChoiceParameter({
+				parameterLongName: '--rerun',
+				description: `Specify what action should be taken when a migration that has already been ${verb} is passed to --name.`,
+				alternatives: ['THROW', 'SKIP', 'ALLOW'],
+				defaultValue: 'THROW',
+			}),
+		};
+	}
+
+	onDefineParameters(): void {
+		this._params = ApplyMigrationsAction._defineParameters(this);
+	}
+
+	private static getVerb(direction: string) {
+		return {
+			up: 'applied',
+			down: 'reverted',
+		}[direction as 'up' | 'down'];
+	}
+
+	async onExecute(): Promise<void> {
+		const {
+			to: { value: to },
+			step: { value: step },
+			name: { values: nameArray },
+			rerun: { value: rerun },
+		} = this._params;
+
+		// string list parameters are always defined. When they're empty it means nothing was passed.
+		const maybeNameArray = nameArray.length > 0 ? nameArray : undefined;
+
+		if (to && maybeNameArray) {
+			throw new Error(`Can't specify 'to' and 'name' together`);
+		}
+
+		if (to && typeof step === 'number') {
+			throw new Error(`Can't specify 'to' and 'step' together`);
+		}
+
+		if (typeof step === 'number' && maybeNameArray) {
+			throw new Error(`Can't specify 'step' and 'name' together`);
+		}
+
+		if (rerun !== 'THROW' && !maybeNameArray) {
+			throw new Error(`Can't specify 'rerun' without 'name'`);
+		}
+
+		const params = {
+			to: to === '0' ? 0 : to,
+			step,
+			migrations: maybeNameArray,
+			rerun,
+		};
+		const actions = {
+			up: async () => this.umzug.up(params as MigrateUpOptions),
+			down: async () => this.umzug.down(params as MigrateDownOptions),
+		};
+		const result = await actions[this.actionName as 'up' | 'down']();
+
+		const verb = ApplyMigrationsAction.getVerb(this.actionName);
+
+		this.umzug.options.logger?.info({ event: this.actionName, message: `${verb} ${result.length} migrations.` });
+	}
+}
+
+export class UpAction extends ApplyMigrationsAction {
+	constructor(umzug: Umzug<{}>) {
```

Comment:
Can `unknown` be used here (and in the other places where you also used `Umzug<{}>`) instead?

```suggestion
	constructor(umzug: Umzug<unknown>) {
```

Also, I thought [ban-types](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/ban-types.md) would not allow you to do this :sweat_smile:

Reply:
It might not, but it's turned off in this repo! I find it both too aggressive and kinda wrong. It turns `{}` into `Record<string, unknown>` even though those aren't really the same things. Anyway, I've made `unknown` the default typearg so these are all just `Umzug` now.